### PR TITLE
Add embeddings

### PR DESCRIPTION
**Pull request by salihadfid1**
This program embeddingsoverIATI does the following:

1. Reads in query and IATI records - and puts them through preprocessing.
2. Unpickles term document matrix with iati-identifiers.
3. Function:  _list_for_wordtovec_ puts all preprocessed records including query and IATI records in a list of list, so it is in a format ready to pass to word2vev
4. Word2Vec function in Gensim is called with the list of list from step 3.
5. The embeddings- vectors generated by word2vec is put into a dataframe (_dfvectors_) - this is done by function _embeddings_in_df._
6. A list is also set up for embeddings that are associated with query, A separate list contains all the words in the query . This is done by function _embeddings_for_qry_
7. For every vector associated with qry , word vectors from iati data- dfvectors are retrieved as long as the cosine similarity between a query vector and word vectors from iati is greater than 0.9.
8. All words in dfvectors that have a cosine similarity greater than 0.9 with words in query vector are put into a list called _similarwords_  which is a list of tuples - each tuple containing the word retrieved from iati records with cosine score.
9. All retrieved word vectors that have similarity score greater than 0.9 are ranked (sorted) in descending order

### CODE REVIEW

**Comment by salihadfid1** [comment]
Please review Add embeddings  2 - latest version

File: python/constants.py [hunk 1]
```diff
@@ -5,6 +5,12 @@
 VECTORIZER_FILENAME = "vectorizer.pkl"
 COSINE_FILENAME = "cosine_df.csv"
 STOPWORDS_FILENAME = "stopwords.txt"
+IATI_IDENTIFIERS_FILENAME = "iati_identifiers_list.pkl"
```
Comment (tqmbow) [inline python/constants.py]: I am not sure we need to run pre-processing separately for embeddings as looks like you're using the same steps as we have been for the tf-idf TDM . In which case can just use the existing constants.py parameter: PROCESSED_RECORDS_FILENAME = "processed_records.csv" and then subset it to just iati-identifiers once read in.

File: python/constants.py [hunk 2]
```diff
@@ -5,6 +5,12 @@
...
+MODEL_NAME = "model.bin"
```
Comment (tqmbow) [inline python/constants.py]: In master will see I've added constants.py of WORD2VECMODEL_FILENAME = "word2vec.model" - can you change it to this?

File: python/preprocessing.py [hunk 3]
```diff
@@ -123,6 +124,14 @@ def preprocess_query_text(query_text):
     return preprocessing_empty_text_remove(df, "description")
 
 
+def preprocess_embeddings(p_df):
```
Comment (tqmbow) [inline python/preprocessing.py]: Is this any different to the preprocessing we are doing currently, which results in processed_records.csv being written out?

File: python/vectorize.py [hunk 4]
```diff
@@ -9,6 +9,7 @@
     WORD_LIST_FILENAME,
     TERM_DOCUMENT_MATRIX_FILENAME,
     VECTORIZER_FILENAME,
+    IATI_IDENTIFIERS_FILENAME,
```
Comment (tqmbow) [inline python/vectorize.py]: As above can we just use PROCESSED_RECORDS_FILENAME ?
Is it necessary to have a list of iati-identifiers written out as a file when should be quick to generate by reading in the CSV specified by PROCESSED_RECORDS_FILENAME?

File: python/constants.py (hunk 2 above)
Comment (nathanbegbie) [reply python/constants.py]: we can sort out the discrepancies in another PR

File: python/constants.py (hunk 1 above)
Comment (nathanbegbie) [reply python/constants.py]: we'll sort this out once it's merged

**Comment by salihadfid1** [comment]
> @salihadfid1 I've done some work on this PR to make sure that the code is formatted and that it is up to date with the master branch. However, please could you remove the preprocess_embeddings function from preprocessing.py as this is replicating existing functionality.

@nathanbegbie  Removed unnecessary code and pushed changes

**Comment by nathanbegbie** [comment]
Thanks @salihadfid1 - we can ignore the codecov issues for now and write tests in the next Pull Request, in order to increase test coverage.

@tqmbow are you happy with the changes?

**Pull request merged**